armi/utils/asciimaps.py

```diff
@@ -40,236 +60,428 @@ class AsciiMap(object):
 
         See Also
         --------
-        armi.reactor.grids : Grid objects that this information can be
-            turned into
+        _updateDimensionsFromAsciiLines : used when reading info from ascii lines
         """
-        self.lattice = {}
-        for row, line in enumerate(reversed(text.strip().splitlines())):
-            line = line.strip()
-            base = self._getRowBase(row)
-            for col, char in enumerate(line.split()):
-                i, j = self._getIndices(base, col)
-                self.lattice[i, j] = char
-
-        self.lattice = {k: v for k, v in self.lattice.items() if v != PLACEHOLDER}
-        return self.lattice
-
-    def writeMap(self, stream):
-        """Write an ascii map from the internal lattice structure to a stream."""
-        colI, rowJ = zip(*sorted(self.lattice.keys(), key=lambda ij: ([ij[1], ij[0]])))
-        line = []
-        for i, j in zip(colI, reversed(rowJ)):
-            if line and i == 0:
-                stream.write(" ".join(line) + "\n")
-                line = []
-            line.append(f"{self.lattice.get((i,j), PLACEHOLDER)}")
-        stream.write(" ".join(line) + "\n")
+        self._ijMax = max(sum(key) for key in self.asciiLabelByIndices)
 
     @staticmethod
-    def _getRowBase(row):
-        raise NotImplementedError()
+    def fromReactor(reactor):
+        """
+        Populate mapping from a reactor in preparation of writing out to ascii.
+        """
 
-    @staticmethod
-    def _getIndices(base, col):
-        raise NotImplementedError()
+    def indicesToAsciiLines(self):
+        """
+        Convert a prepared asciiLabelByIndices to ascii lines and offsets.
+
+        This is used when you have i,j/specifier data and want to create a ascii map from it
+        as opposed to reading a ascii map from a stream.
+
+        As long as the map knows how to convert lineNum and colNums into ij indices, this
+        is universal. In some implementations, this operation is in a different
+        method for efficiency.
+        """
+        self._updateDimensionsFromData()
+        self.asciiLines = []
+        for lineNum in reversed(range(self._asciiMaxLine)):
+            line = []
+            for colNum in range(self._asciiMaxCol):
+                ij = self._getIJFromColRow(colNum, lineNum)
+                # convert to string and strip any whitespace in thing we're representing
+                line.append(
+                    str(self.asciiLabelByIndices.get(ij, PLACEHOLDER)).replace(" ", "")
+                )
+            self.asciiLines.append(line)
+
+        # clean data
+        noDataLinesYet = True  # handle all-placeholder rows
+        newLines = []
+        for line in self.asciiLines:
+            if re.search("^[-]+$", "".join(line)) and noDataLinesYet:
+                continue
+            else:
+                noDataLinesYet = False
+                newLine = self._removeTrailingPlaceholders(line)
```

Comment:
another way to see if a row is just placeholders. Could use this above, though more expensive i suppose